Updated, Multi-Stage Dockerfile

seunomonije opened the pull request:
**Updated Dockerfile on Python 3.8**

Changes:
 - Cleaned apt cache when installing dependencies. See [Dockerfile Best Practices](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/) for more information
 - Cleaned up installing package from multiple requirements files.
 - Followed [multi-stage build practices](https://pythonspeed.com/articles/multi-stage-docker-python). 

Notes:
 - New image size is 1.14GB
 - Built the image using `docker build -t cirq .`

seunomonije commented:
> @googlebot I signed it!

seunomonije commented:
> We found a Contributor License Agreement for you (the sender of this pull request), but were unable to find agreements for all the commit author(s) or Co-authors. If you authored these, maybe you used a different email address in the git commits than was used to sign the CLA ([login here](https://cla.developers.google.com/) to double check)? If these were authored by someone else, then they will need to sign a CLA as well, and confirm that they're okay with these being contributed to Google.
> In order to pass this check, please resolve this problem and then comment `@googlebot I fixed it.`. If the bot doesn't comment, it means it doesn't think anything has changed.
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fquantumlib%2FCirq%2Fpull%2F3553) for more info**.

@googlebot I fixed it.

seunomonije commented:
> We found a Contributor License Agreement for you (the sender of this pull request), but were unable to find agreements for all the commit author(s) or Co-authors. If you authored these, maybe you used a different email address in the git commits than was used to sign the CLA ([login here](https://cla.developers.google.com/) to double check)? If these were authored by someone else, then they will need to sign a CLA as well, and confirm that they're okay with these being contributed to Google.
> In order to pass this check, please resolve this problem and then comment `@googlebot I fixed it.`. If the bot doesn't comment, it means it doesn't think anything has changed.
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fquantumlib%2FCirq%2Fpull%2F3553) for more info**.

@googlebot I fixed it.

balopat commented:
Hey @seunomonije  - your commits are by oseun22 at gmail dot com, but maybe you're using a different email address to sign the CLA?

seunomonije commented:
> We found a Contributor License Agreement for you (the sender of this pull request), but were unable to find agreements for all the commit author(s) or Co-authors. If you authored these, maybe you used a different email address in the git commits than was used to sign the CLA ([login here](https://cla.developers.google.com/) to double check)? If these were authored by someone else, then they will need to sign a CLA as well, and confirm that they're okay with these being contributed to Google.
> In order to pass this check, please resolve this problem and then comment `@googlebot I fixed it.`. If the bot doesn't comment, it means it doesn't think anything has changed.
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fquantumlib%2FCirq%2Fpull%2F3553) for more info**.

@googlebot I fixed it.

seunomonije commented:
> Hey @seunomonije - your commits are by oseun22 at gmail dot com, but maybe you're using a different email address to sign the CLA?

Let me and sign it with oseun22@gmail.com. I thought the commits were according to my GitHub email, which is seunomonije@icloud.com.

balopat commented on dev_tools/Dockerfile:
```diff
@@ -11,14 +21,24 @@ ENV LANGUAGE en_US:en
+     && ln -s /usr/bin/python3 /usr/bin/python
 
 # Get a current copy of Cirq from Github.
 RUN git clone https://github.com/quantumlib/Cirq
```
instead of cloning, we should just `COPY` the current folder

seunomonije commented:
> > Hey @seunomonije - your commits are by oseun22 at gmail dot com, but maybe you're using a different email address to sign the CLA?
> 
> Let me and sign it with [oseun22@gmail.com](mailto:oseun22@gmail.com). I thought the commits were according to my GitHub email, which is [seunomonije@icloud.com](mailto:seunomonije@icloud.com).

@balopat I added the CLA to oseun22@gmail.com

balopat commented on dev_tools/Dockerfile:
```diff
@@ -1,8 +1,18 @@
+# rm -rf /var/lib/apt/lists/* cleans up apt cache. See https://docs.docker.com/develop/develop-images/dockerfile_best-practices/
 RUN DEBIAN_FRONTEND=noninteractive apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y \
-     python3-pip python3-tk texlive-latex-base latexmk git emacs vim locales
+     python3-pip \
```
do we still need pip? python 3.7-slim doesn't come with it?

seunomonije replied:
Good call.

seunomonije replied:
Turns out that we need pip bc it includes gcc, which is required to install the immutables package (dependency of pyquil) and cytoolz

balopat replied:
I think this is fine for now.

balopat commented on dev_tools/Dockerfile:
```diff
@@ -11,14 +21,24 @@ ENV LANGUAGE en_US:en
+
 # Install the needed Python libraries.
-RUN pip3 install -r Cirq/requirements.txt -r Cirq/cirq/contrib/contrib-requirements.txt -r Cirq/dev_tools/conf/pip-list-dev-tools.txt
+RUN pip3 install --user .
```
I'm surprised that this works - `pip install --user .` would install the current directory, but Cirq is in the `./Cirq` directory...

seunomonije replied:
Looks like I made an error here when cleaning before pushing to github. Fixing it rn

balopat commented on dev_tools/Dockerfile:
```diff
@@ -11,14 +21,24 @@ ENV LANGUAGE en_US:en
 # Get a current copy of Cirq from Github.
 RUN git clone https://github.com/quantumlib/Cirq
 
+COPY requirements.txt .
```
This is good, you're already adding things from the local folders here.

balopat commented on dev_tools/Dockerfile:
```diff
@@ -11,14 +21,24 @@ ENV LANGUAGE en_US:en
 RUN git clone https://github.com/quantumlib/Cirq
 
+COPY requirements.txt .
+COPY Cirq/cirq/contrib/contrib-requirements.txt .
```
But these two should change and you could just use `./cirq/contrib/contrib-requirements` and `./dev_tools/conf/pip-list-dev-tools.txt`.

balopat commented on dev_tools/Dockerfile:
```diff
@@ -11,14 +21,24 @@ ENV LANGUAGE en_US:en
+COPY Cirq/cirq/contrib/contrib-requirements.txt .
+COPY Cirq/dev_tools/conf/pip-list-dev-tools.txt .
+
+RUN cat requirements.txt contrib-requirements.txt pip-list-dev-tools.txt >> dependencies.txt \
```
no need for separate file creation, you can pass in multiple `-r <file>` args to pip, like `pip install -r req1.txt -r req2.txt`

balopat reviewed (changes requested):
Thanks for raising @seunomonije! This is pointing towards the right direction. Added a bunch of comments!

Also, it would be great to add a `test docker` job in CI, it would 
- build the dockerfile 
- run a simple tests: e.g. `docker run cirq-docker python -c "import cirq; print(cirq.Circuit(cirq.H(cirq.NamedQubit(a))))" and would test what the output is

balopat commented:
> * Is there any reason for multiple Dockerfiles? There is one in the root directory, and one in `\dev_tools`

That one is for API docs testing only. 

> > Hey @seunomonije - your commits are by oseun22 at gmail dot com, but maybe you're using a different email address to sign the CLA?
> 
> Let me and sign it with [oseun22@gmail.com](mailto:oseun22@gmail.com). I thought the commits were according to my GitHub email, which is [seunomonije@icloud.com](mailto:seunomonije@icloud.com).

It depends on what your local git settting is - `git config user.email` should show you that. See https://docs.github.com/en/free-pro-team@latest/github/setting-up-and-managing-your-github-user-account/setting-your-commit-email-address#setting-your-commit-email-address-in-git for more info.

balopat commented on dev_tools/Dockerfile:
```diff
@@ -1,8 +1,18 @@
-FROM ubuntu
+FROM python:3.7-slim AS compile-image
```
Let's go with 3.8, the upper boundary of our supported python versions.

balopat commented on dev_tools/Dockerfile:
```diff
@@ -11,14 +21,24 @@ ENV LANGUAGE en_US:en
 
-WORKDIR /Cirq
+FROM python:3.7-slim AS build-image
+COPY --from=compile-image /root/.local /root/.local
```
thinking about this - it might be problematic to do the multistage build this way because the final image will be missing all the visualization libraries that we apt-get installed :/ Can you try to run an example or a notebook that uses matplotlib? You should be able to run `docker run -p 8888:8888 cirq-image jupyter notebook --port=8888 --no-browser --ip=0.0.0.0` and then connect to it via http://localhost:8888

seunomonije replied:
I see what you're saying. 
I'm getting this error:
<details>
<summary>Error</summary>
<p>
docker: Error response from daemon: OCI runtime create failed: container_linux.go:349: starting container process caused "exec: \"jupyter\": executable file not found in $PATH": unknown.
</p>
</details>

When running the file. What is the correct way to handle this?

seunomonije replied:
Figured it out. Jupyter is not installed by default on python base image.

seunomonije replied:
@balopat I made some changes and I was actually just able to run a simple example with matplotlib successfully in my browser. I then ran the notebook at `docs/tutorials/basics.ipynb` and everything worked as well.

balopat replied:
Hey @seunomonije - nice! Thank you for pushing through with this, yes jupyter was missing. 
No need to open a fresh PR, you can just convert this one into a non-draft PR - it's better to have all the conversation on the same PR, so when questions arise we can come back to these decisions.

seunomonije commented:
Marking as draft until I get to it later this week.

balopat commented on dev_tools/Dockerfile:
```diff
@@ -1,8 +1,18 @@
+     texlive-latex-base \
+     latexmk \
+     git \
+     emacs \
```
let's remove emacs and vim, it is not our responsibility to bundle these tools in my opinion.

seunomonije commented:
The only current issue is the need for the `--allow-root` flag is needed to run the jupyter notebook.

balopat commented:
> **Notes:**
> 
> The `--allow-root` flag is needed to run the jupyter notebook.
> 
> There seem to be unnamed images when building the cirq image. See picture. Should I include in a command when running the image to clean?
> <img alt="Screen Shot 2021-01-05 at 7 13 18 PM" width="397" src="https://user-images.githubusercontent.com/17647506/103717030-1321de80-4f8a-11eb-93e6-067c42c1203e.png">

Those are pretty normal. The previous images stay untagged and/or the layers from the multilayer build. You can clean those up with `docker rmi <imageid>`.

balopat commented on dev_tools/Dockerfile:
```diff
@@ -1,8 +1,16 @@
+FROM python:3.8-slim AS compile-image
 
 # Install dependencies.
+# rm -rf /var/lib/apt/lists/* cleans up apt cache. See https://docs.docker.com/develop/develop-images/dockerfile_best-practices/
```
why the comment? we can add this after the apt-get install command

balopat commented on dev_tools/Dockerfile:
```diff
@@ -11,14 +19,30 @@ ENV LANGUAGE en_US:en
-# Install the needed Python libraries.
-RUN pip3 install -r Cirq/requirements.txt -r Cirq/cirq/contrib/contrib-requirements.txt -r Cirq/dev_tools/conf/pip-list-dev-tools.txt
+# Copy current folder instead of cloning.
+COPY ./ .
```
```suggestion
COPY . .
```
